server/src/frameworks/Hardhat/HardhatProject.ts

```diff
@@ -0,0 +1,298 @@
+
+  private logger: Logger;
+
+  private name: string;
+
+  constructor(
+    serverState: ServerState,
+    basePath: string,
+    public configPath: string
+  ) {
+    super(serverState, basePath);
+    this.logger = _.cloneDeep(serverState.logger);
+    this.name = path.basename(basePath);
+    this.logger.tag = `${this.name}`;
+  }
+
+  public id(): string {
+    return this.configPath;
+  }
+
+  public frameworkName(): string {
+    return "Hardhat";
+  }
+
+  public async initialize(): Promise<void> {
+    return new Promise((resolve, _reject) => {
+      this._onInitialized = resolve;
+
+      // Fork WorkerProcess as child process
+      this.workerProcess = fork(
+        path.resolve(__dirname, "worker/WorkerProcess.js"),
+        {
+          cwd: this.basePath,
+          detached: true,
+          execArgv: [],
+        }
+      );
+      this.workerStatus = WorkerStatus.INITIALIZING;
+
+      this.workerProcess.on("message", async (message: Message) => {
+        await this._handleMessage(message);
+      });
+
+      this.workerProcess.on("error", (err) => {
+        this.logger.error(err);
+      });
+
+      this.workerProcess.on("exit", async (code) => {
+        this.logger.error(`Child process exited: ${code}`);
+        this.workerStatus = WorkerStatus.ERRORED;
+        this._onInitialized();
+      });
+    });
+  }
+
+  public async fileBelongs(sourceURI: string): Promise<boolean> {
+    const workerPromise = new Promise((resolve, reject) => {
+      this._checkWorkerExists();
+      this._checkWorkerNotInitializing();
+
+      if (this.workerStatus === WorkerStatus.RUNNING) {
+        // HRE was loaded successfully. Delegate to the worker that will use the configured sources path
+        const requestId = this._prepareRequest(resolve, reject);
+
+        this.workerProcess!.send(new FileBelongsRequest(requestId, sourceURI));
+      } else {
+        // HRE could not be loaded. Claim ownership of all solidity files under root folder
+        // This is to avoid potential hardhat-owned contracts being assigned to i.e. projectless
+        resolve(directoryContains(this.basePath, sourceURI));
+      }
+    });
+
+    return Promise.race([
+      workerPromise,
+      this._requestTimeout("fileBelongs"),
+    ]) as Promise<boolean>;
+  }
+
+  public async buildCompilation(
+    sourceUri: string,
+    openDocuments: OpenDocuments
+  ): Promise<CompilationDetails> {
+    const workerPromise = new Promise((resolve, reject) => {
+      this._checkWorkerExists();
+      this._checkWorkerNotInitializing();
+
+      if (this.workerStatus === WorkerStatus.ERRORED) {
+        throw new Error(this.workerLoadFailureReason);
+      }
+
+      const requestId = this._prepareRequest(resolve, reject);
+
+      this.workerProcess!.send(
+        new BuildCompilationRequest(requestId, sourceUri, openDocuments)
+      );
+    });
+
+    return Promise.race([
+      workerPromise,
+      this._requestTimeout("buildCompilation"),
```

Comment:
I don't think it's needed to reject the worker promise, since the parent promise (the race one) encapsulates it. If a timeout happens, the `buildCompilation` function is rejected with the timeout error. Eventually it gets shown in the UI in a status item (default error behavior).